### ISSUE TITLE
browser.waitUntil not passing this context to condition function (ES6 page object)

### ISSUE THREAD

**Issue by mherodev** [webdriverio/webdriverio#2559]
## The problem

When trying to call my page object's methods which use `this` or `browser` inside `browser.waitUntil`, I am seeing errors where things that are normally defined are not defined, e.g.

`this.firstTheater.$ is not a function`
`this.theaterFavoriteButton.getAttribute is not a function`

where I'm calling

```javascript
    browser.waitUntil(
      () => theaterSelection.isTheaterFavorite(),
      5000,
      'expected text to be different after 5s'
    );
```

where my page object is defined like

```javascript
  get firstTheater() {
    return $('.showtimes .theater:first-child');
  }

  get theaterFavoriteButton() {
    return $('.showtimes .theater:first-child .theater__follow-icon');
  }

  isTheaterFavorite() {
    const isFavorite = browser.getAttribute(
      '.showtimes .theater:first-child .theater__follow-icon',
      'data-is-favorite'
    );
    return isFavorite === 'true';
  }

 // or, optimally...

  isTheaterFavorite() {
    const isFavorite = this.theaterFavoriteButton.getAttribute(
      'data-is-favorite'
    );
    return isFavorite === 'true';
  }
```

This does not change if I convert the callback from an arrow function to a regular one. I can work around this however by not using the Page Object at all, but that's killing the usefulness of the pattern.

## Environment

* WebdriverIO version: 4.9.8
* Node.js version: v8.8.1
* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html): testrunner
* if wdio testrunner, running synchronous or asynchronous tests: synchronouse
* Additional wdio packages used (if applicable):

**Comment by christian-bromann** [comment]
@mherodev this should actually work. Not sure why not since there are no logs available. You definitely should be able to use this within your PO. Closing until this is actually turns out to be a bug.

**Comment by mherodev** [comment]
I still am not certain where to get the logs you need. The logs saved by webdriverio don’t appear to have anything useful in them.

**Comment by christian-bromann** [comment]
@mherodev it depends how you run a browser driver .. if you use a service there is an option to store logs into a file

**Comment by mherodev** [comment]
Ah, I think I see where they're being written in Sauce Labs. I'll run the broken implementation shortly and get those logs. 👌

**Comment by mherodev** [comment]
Huh, I take it back, none of these logs seem to be useful. Here they are, nonetheless:

The test
```javascript
  test.only('Favoriting a theater', () => {
    browser.waitUntil(
      () => theaterSelection.isTheaterFavorite(),
      10000,
      'expected favorite to be different after 10s'
    );
  });
```

local console output
```
TypeError: this.theaterFavoriteButton.getAttribute is not a function
```

SauceLabs log.json
```
[
  {
    "screenshot": 0,
    "between_commands": null,
    "suggestion_values": [],
    "request": {
      "desiredCapabilities": {
        "rotatable": true,
        "javascriptEnabled": true,
        "locationContextEnabled": true,
        "chromeOptions": {
          "binary": "/Volumes/Sauce/Chrome/Chrome 63.app/Contents/MacOS/Google Chrome",
          "args": [
            "start-maximized",
            "disable-webgl",
            "blacklist-webgl",
            "blacklist-accelerated-compositing",
            "disable-accelerated-2d-canvas",
            "disable-accelerated-compositing",
            "disable-accelerated-layers",
            "disable-accelerated-plugins",
            "disable-accelerated-video",
            "disable-accelerated-video-decode",
            "disable-gpu",
            "disable-infobars",
            "test-type"
          ]
        },
        "browserName": "chrome",
        "loggingPrefs": {
          "driver": "ALL",
          "browser": "ALL"
        },
        "handlesAlerts": true,
        "requestOrigins": {
          "url": "http://webdriver.io",
          "version": "4.9.8",
          "name": "webdriverio"
        }
      }
    },
    "HTTPStatus": 200,
    "result": {
      "takesScreenshot": true,
      "acceptSslCerts": true,
      "networkConnectionEnabled": false,
      "mobileEmulationEnabled": false,
      "unexpectedAlertBehaviour": "",
      "applicationCacheEnabled": false,
      "locationContextEnabled": true,
      "rotatable": false,
      "chrome": {
        "chromedriverVersion": "2.34.522932 (4140ab217e1ca1bec0c4b4d1b148f3361eb3a03e)",
        "userDataDir": "/var/folders/5n/qvrkkll11sg_5xwc5_qcx4c40000kr/T/.org.chromium.Chromium.h0ma34"
      },
      "hasTouchScreen": false,
      "platform": "Mac OS X",
      "version": "63.0.3239.84",
      "nativeEvents": true,
      "handlesAlerts": true,
      "takesHeapSnapshot": true,
      "javascriptEnabled": true,
      "databaseEnabled": false,
      "browserName": "chrome",
      "webStorageEnabled": true,
      "browserConnectionEnabled": false,
      "cssSelectorsEnabled": true,
      "setWindowRect": true,
      "pageLoadStrategy": "normal"
    },
    "suggestion": null,
    "duration": 2.0462210178375244,
    "path": "/session",
    "method": "POST",
    "statusCode": 0,
    "timelineSeconds": 0
  },
  {
    "screenshot": null,
    "between_commands": 0.7354068756103516,
    "suggestion_values": [],
    "request": {
      "width": 960,
      "height": 1200
    },
    "HTTPStatus": 200,
    "result": null,
    "suggestion": null,
    "duration": 0.3262500762939453,
    "path": "window/current/size",
    "method": "POST",
    "statusCode": 0,
    "timelineSeconds": 2.0462210178375244
  },
  {
    "screenshot": 1,
    "between_commands": 0.13772892951965332,
    "suggestion_values": [],
    "request": {
      "url": "https://www.mysite.com/mypath"
    },
    "HTTPStatus": 200,
    "result": null,
    "suggestion": null,
    "duration": 10.080379009246826,
    "path": "url",
    "method": "POST",
    "statusCode": 0,
    "timelineSeconds": 3.1078779697418213
  },
  {
    "screenshot": null,
    "between_commands": 0.15726709365844727,
    "suggestion_values": [],
    "request": {
      "using": "css selector",
      "value": ".showtimes .theater:first-child .theater__follow-icon"
    },
    "HTTPStatus": 200,
    "result": {
      "ELEMENT": "0.21356838979194714-1"
    },
    "suggestion": null,
    "duration": 0.025589942932128906,
    "path": "element",
    "method": "POST",
    "statusCode": 0,
    "timelineSeconds": 13.3259859085083
  },
  {
    "screenshot": 2,
    "between_commands": 0.22828388214111328,
    "suggestion_values": [],
    "request": {},
    "HTTPStatus": 200,
    "result": "",
    "suggestion": null,
    "duration": 0.00010919570922851562,
    "path": "/session/ad165992e80949adb8c10a9a2411965b",
    "method": "DELETE",
    "statusCode": 0,
    "timelineSeconds": 13.508842945098877
  }
]
```

SauceLabs selenium-server.log
```
[0.349][INFO]: COMMAND InitSession {
   "desiredCapabilities": {
      "browserName": "chrome",
      "chromeOptions": {
         "args": [ "start-maximized", "disable-webgl", "blacklist-webgl", "blacklist-accelerated-compositing", "disable-accelerated-2d-canvas", "disable-accelerated-compositing", "disable-accelerated-layers", "disable-accelerated-plugins", "disable-accelerated-video", "disable-accelerated-video-decode", "disable-gpu", "disable-infobars", "test-type" ],
         "binary": "/Volumes/Sauce/Chrome/Chrome 63.app/Contents/MacOS/Google Chrome"
      },
      "handlesAlerts": true,
      "javascriptEnabled": true,
      "locationContextEnabled": true,
      "loggingPrefs": {
         "browser": "ALL",
         "driver": "ALL"
      },
      "requestOrigins": {
         "name": "webdriverio",
         "url": "http://webdriver.io",
         "version": "4.9.8"
      },
      "rotatable": true,
      "webdriver.remote.quietExceptions": false
   }
}
[0.350][INFO]: Populating Preferences file: {
   "alternate_error_pages": {
      "enabled": false
   },
   "autofill": {
      "enabled": false
   },
   "browser": {
      "check_default_browser": false
   },
   "distribution": {
      "import_bookmarks": false,
      "import_history": false,
      "import_search_engine": false,
      "make_chrome_default_for_user": false,
      "skip_first_run_ui": true
   },
   "dns_prefetching": {
      "enabled": false
   },
   "profile": {
      "content_settings": {
         "pattern_pairs": {
            "https://*,*": {
               "media-stream": {
                  "audio": "Default",
                  "video": "Default"
               }
            }
         }
      },
      "default_content_setting_values": {
         "geolocation": 1
      },
      "default_content_settings": {
         "geolocation": 1,
         "mouselock": 1,
         "notifications": 1,
         "popups": 1,
         "ppapi-broker": 1
      },
      "password_manager_enabled": false
   },
   "safebrowsing": {
      "enabled": false
   },
   "search": {
      "suggest_enabled": false
   },
   "translate": {
      "enabled": false
   }
}
[0.350][INFO]: Populating Local State file: {
   "background_mode": {
      "enabled": false
   },
   "ssl": {
      "rev_checking": {
         "enabled": false
      }
   }
}
[0.352][INFO]: Launching chrome: /Volumes/Sauce/Chrome/Chrome 63.app/Contents/MacOS/Google Chrome --blacklist-accelerated-compositing --blacklist-webgl --disable-accelerated-2d-canvas --disable-accelerated-compositing --disable-accelerated-layers --disable-accelerated-plugins --disable-accelerated-video --disable-accelerated-video-decode --disable-background-networking --disable-client-side-phishing-detection --disable-default-apps --disable-gpu --disable-hang-monitor --disable-infobars --disable-popup-blocking --disable-prompt-on-repost --disable-sync --disable-web-resources --disable-webgl --enable-automation --enable-logging --force-fieldtrials=SiteIsolationExtensions/Control --ignore-certificate-errors --load-extension=/var/folders/5n/qvrkkll11sg_5xwc5_qcx4c40000kr/T/.org.chromium.Chromium.VoJ3QM/internal --log-level=0 --metrics-recording-only --no-first-run --password-store=basic --remote-debugging-port=12923 --start-maximized --test-type --use-mock-keychain --user-data-dir=/var/folders/5n/qvrkkll11sg_5xwc5_qcx4c40000kr/T/.org.chromium.Chromium.h0ma34 data:,
[2.325][INFO]: resolved localhost to ["::1","127.0.0.1"]
[2.386][INFO]: RESPONSE InitSession {
   "acceptSslCerts": true,
   "applicationCacheEnabled": false,
   "browserConnectionEnabled": false,
   "browserName": "chrome",
   "chrome": {
      "chromedriverVersion": "2.34.522932 (4140ab217e1ca1bec0c4b4d1b148f3361eb3a03e)",
      "userDataDir": "/var/folders/5n/qvrkkll11sg_5xwc5_qcx4c40000kr/T/.org.chromium.Chromium.h0ma34"
   },
   "cssSelectorsEnabled": true,
   "databaseEnabled": false,
   "handlesAlerts": true,
   "hasTouchScreen": false,
   "javascriptEnabled": true,
   "locationContextEnabled": true,
   "mobileEmulationEnabled": false,
   "nativeEvents": true,
   "networkConnectionEnabled": false,
   "pageLoadStrategy": "normal",
   "platform": "Mac OS X",
   "rotatable": false,
   "setWindowRect": true,
   "takesHeapSnapshot": true,
   "~~~": "..."
}
[2.558][INFO]: COMMAND ExecuteScript {
   "args": [  ],
   "script": "return screen.width"
}
[2.558][INFO]: Waiting for pending navigations...
[2.640][INFO]: Done waiting for pending navigations. Status: ok
[2.648][INFO]: Waiting for pending navigations...
[2.649][INFO]: Done waiting for pending navigations. Status: ok
[2.649][INFO]: RESPONSE ExecuteScript 1024
[2.652][INFO]: COMMAND ExecuteScript {
   "args": [  ],
   "script": "return screen.height"
}
[2.652][INFO]: Waiting for pending navigations...
[2.653][INFO]: Done waiting for pending navigations. Status: ok
[2.662][INFO]: Waiting for pending navigations...
[2.664][INFO]: Done waiting for pending navigations. Status: ok
[2.664][INFO]: RESPONSE ExecuteScript 768
[2.667][INFO]: COMMAND SetWindowPosition {
   "windowHandle": "current",
   "x": 0,
   "y": 0
}
[2.668][INFO]: resolved localhost to ["::1","127.0.0.1"]
[2.796][INFO]: RESPONSE SetWindowPosition
[2.799][INFO]: COMMAND SetWindowSize {
   "height": 768,
   "width": 1024,
   "windowHandle": "current"
}
[2.965][INFO]: RESPONSE SetWindowSize
[2.968][INFO]: COMMAND MaximizeWindow {
   "windowHandle": "current"
}
[2.969][INFO]: RESPONSE MaximizeWindow
[2.972][INFO]: COMMAND GetWindows {

}
[2.973][INFO]: RESPONSE GetWindows [ "CDwindow-(11F8A653429FE0C5E1530ECE643DA4D8)" ]
[3.125][INFO]: COMMAND SetWindowSize {
   "height": 1200,
   "width": 960,
   "windowHandle": "current"
}
[3.448][INFO]: RESPONSE SetWindowSize
[3.589][INFO]: COMMAND Navigate {
   "url": "https://www.mysite.com/mypath"
}
[3.589][INFO]: Waiting for pending navigations...
[3.591][INFO]: Done waiting for pending navigations. Status: ok
[4.601][INFO]: Waiting for pending navigations...
[13.666][INFO]: Done waiting for pending navigations. Status: ok
[13.666][INFO]: RESPONSE Navigate
[13.826][INFO]: COMMAND FindElement {
   "using": "css selector",
   "value": ".showtimes .theater:first-child .theater__follow-icon"
}
[13.826][INFO]: Waiting for pending navigations...
[13.830][INFO]: Done waiting for pending navigations. Status: ok
[13.843][INFO]: Waiting for pending navigations...
[13.849][INFO]: Done waiting for pending navigations. Status: ok
[13.849][INFO]: RESPONSE FindElement {
   "ELEMENT": "0.21356838979194714-1"
}

```

**Comment by mherodev** [comment]
I'm seeing the root of this issue in other places, e.g.

```javascript
  get movieTitle() {
    return this.firstMovie.$('.fd-movie__title a');
  }
```

This call appears to work fine for `waitForVisible`; however, if I call

```javascript
myPageObject.firstMovie.getHtml(false)
```

I am seeing an HTML array for all movie titles on the page, instead of only the single title scoped to the `firstMovie` parent element.

I can fix this by defining `movieTitle` as follows:

``` javascript
  get movieTitle() {
    return $(`${this.firstMovie.selector} .fd-movie__title a`);
  }
```

There's something funky with how the element chaining is done in wdio, but in hindsight I wish I'd have avoided using it. `browser.whatever` is much more reliable
